### ISSUE TITLE
2.2.11 added a bug where cross package dependencies are no longer compiled correctly

### ISSUE THREAD

**Issue by ldriscoll** [immutables/immutables#421]
We have immutables that get injected as a generic in a super interface of another immutable, and these are in different packages.  Since 2.2.11 this stopped compiling.

Here is an example
```java
package a;
public interface Inner<T> {
  T getStart();
  T getFinish();
}

package a;
public interface OuterFields extends Inner<ImmutableFoo> {
 }

package a;
@Value.Immutable
public interface Outer extends OuterFields {
}

package b;
@Value.Immutable
public interface Foo {
  String getName();
} 

```

**Comment by elucash** [comment]
It's almost impossible to provide correct handling for these without rewriting half of javac. Frankly, I don't have a clue how it worked previously, so I rather implemented additional scanning of inheritance hierarchy. If there will be same named yet-to-be-generated classes from different packages mixed in the single inheritance hierarchy, this will be broken again, so, please avoid it ;)

**Comment by ldriscoll** [comment]
This is very concerning to me, surely immutables code generation should be happening at the generate-sources phase, and more importantly, should not be trying to resolve compiled classes when generating source code.  It would appear that a change in 2.2.11 is causing the code generation to expect other classes to not only be generated first, but be compiled.  

Compilation errors, should be dealt with purely by the compiler.  I'm a little terrified that we're now trying to resolve compiled code at code generation time.

**Comment by elucash** [comment]
We can only "overcome" this if we fully analyze a whole hierarchy including analysis of the source code to preserve type structure which might be lost once compiler may flag it as unresolved, we need to compute all subtyping and type variable substitutions across whole hierarchy to exactly know from which source file which value comes and when there are star-imports..., well, we're busted anyway. It's easy to say that Javac and ECJ are buggy (which they are, as most software), but I, honestly, could not expect it to resolve all the cases as there are no enough information about yet-to-be-generated types. The whole idea is wrong (including annotation processing itself), I'm sorry, not-yet-generated type are second class citizens, so it's better to avoid them, at least in signatures/api.
Have you considered moving to style where generated types have suffixes, but not exposed directly, as illustrated at the top of a [README.md](https://github.com/immutables/immutables/blob/master/README.md) ? This is probably my only recommendation if you've got to use really complex generics/subtyping/codegeneration combinations.

**Comment by elucash** [comment]
released in 2.2.12
